client/timeout/timeout.go

```diff
@@ -0,0 +1,32 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package timeout enforces a maximum timeout on all outgoing rpcs.
+package timeout
+
+import (
+	"context"
+	"time"
+
+	"google.golang.org/grpc"
+)
+
+// UnaryClientInterceptor returns a new maximum timout client interceptor.
```

Comment:
"maximum timeout" makes it sound like it will only make timeouts smaller if they are larger than maxTimeout. If maxTimeout is 10s and an RPC is made with a 1s deadline, it looks like this will upgrade it to a 10s timeout?

(also there's a typo in timeout)

Reply:
That's correct and the intended behavior.  Implemented by `WithTimeout` when ctx already has a timeout.